### PR TITLE
Add documentation to modin/pandas/base.py

### LINKED ISSUE #2942: Document modin.pandas/base.py
This should include docstrings and purposes of `modin.pandas/base.py`.

### PR DESCRIPTION
**Pull request by gshimansky**
## What do these changes do?

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` 
- [x] Resolves #2942 
- [x] tests added and passing

### CODE REVIEW

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:155]:
```suggestion
        pandas.DataFrame or pandas.Series
            pandas dataset with `num_rows` or fewer rows and `num_cols` or fewer columns.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:238]:
```suggestion
        axis : {None, 0, 1}
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:239]: could you please specify which value corresponds to which axis?

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:254]:
```suggestion
        modin.pandas.BasePandasDataset
            Other frame if it is determined to be valid.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:359]:
```suggestion
        modin.pandas.BasePandasDataset
            Result of binary operation.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:406]:
```suggestion
        object
            Result of operation.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:489]:
```suggestion
        int
            0 or 1 - axis index in the array of axes stored in the dataframe.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:507]:
```suggestion
        modin.pandas.BasePandasDataset
            Constructed object.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:533]:
```suggestion
        pandas.Index
            The union of all indexes across the partitions.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:577]: Does this generate a cross-link to `aggregate` function?

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:610]:
```suggestion
        object
            Function result.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:635]:
```suggestion
        list
            Either a one-element list that contains `dtype` if object denotes a Series
            or a list that contains `dtypes` if object denotes a DataFrame.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:1556]:
```suggestion
        axis : int or str
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:1561]:
```suggestion
        level : int or str
            If specified `axis` is a MultiIndex, apply method along a particular
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:1563]:
```suggestion
        numeric_only : bool, optional
```
I think `optional` suits better when default value is `None` in almost any case

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:1572]:
```suggestion
        scalar, Series or DataFrame
            `scalar` - self is Series and level is not specified.
            `Series` - self is Series and level is specified, or
                 self is DataFrame and level is not specified.
            `DataFrame` - self is DataFrame and level is specified.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:2818]:
```suggestion
        key : location or index-based slice
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:2820]:
```suggestion
        value : object
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:2833]:
```suggestion
        key : location or index-based slice
            Key that points to rows to retrieve.
```

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:2838]:
```suggestion
        modin.pandas.BasePandasDataset
            Selected rows.
```

**Review by vnlitvinov — COMMENTED** [review]
Overall looks good, mostly minor corrections in indentation.

File: modin/pandas/base.py
Comment (gshimansky) [reply modin/pandas/base.py]: There is a problem with `doc_checker.py` now. For low case `pandas` I get an error
```
ERROR:modin.pandas.base.BasePandasDataset._build_repr_df:135:RT04:Return value description should start with a capital letter
```
and for capital `Pandas` it gives error
```
ERROR:modin.pandas.base.BasePandasDataset._build_repr_df:135:MD02:Spelling error in line: 19, found: 'Pandas', reference: 'pandas'
```

Comment (anmyachev) [reply modin/pandas/base.py]: Just do not write the word pandas, this can be understood by the type that is located on the line above.

File: modin/utils.py
Comment (anmyachev) [inline modin/utils.py:115]:
```suggestion
        and "`pandas API documentation for " not in target_doc
```

File: modin/utils.py
Comment (anmyachev) [inline modin/utils.py:123]:
```suggestion
        doc += f"\n\n{' ' * _get_indent(doc)}See `pandas API documentation for {token} <{url}>`_ for more."
```

File: modin/pandas/base.py
Comment (vnlitvinov) [reply modin/pandas/base.py]: @anmyachev I believe we can fix that in #3001, too.

File: modin/pandas/base.py
Comment (vnlitvinov) [inline modin/pandas/base.py:241]:
```suggestion
            validation is done along `columns` of `other` frame.
```
typo?

File: modin/pandas/base.py
Comment (gshimansky) [reply modin/pandas/base.py]: Ok thanks. `# noqa: MD02` helped to fix this warning.

File: modin/pandas/base.py
Comment (gshimansky) [reply modin/pandas/base.py:241]: yes

**Review by vnlitvinov — DISMISSED** [review]
LGTM!

**Review by anmyachev — APPROVED** [review]

File: docs/developer/pandas/base.rst
Comment (vnlitvinov) [inline docs/developer/pandas/base.rst:1]: we should probably move it under `docs/flow/modin` from `docs/developer`?

**Review by anmyachev — CHANGES_REQUESTED** [review]
We should also enable numpydoc check for `modin/pandas/base.py`

**Comment by gshimansky** [comment]
> We should also enable numpydoc check for `modin/pandas/base.py`

Ok I enabled it but it fails. We need to merge fixed `doc_checker` to make it work.

File: .github/workflows/ci.yml
Comment (anmyachev) [inline .github/workflows/ci.yml:55]:
```suggestion
      - run: python scripts/doc_checker.py --add-ignore=D101,D102,D105 modin/pandas/base.py
```

**Review by anmyachev — DISMISSED** [review]
LGTM! Just revert `CI.yml` changes.

**Review by vnlitvinov — DISMISSED** [review]
LGTM assuming CI passes.

**Comment by vnlitvinov** [comment]
> LGTM! Just revert `CI.yml` changes.

We should be able to enable `numpydoc` against `base.py` after #3001 lands (or in it itself).

**Review by anmyachev — APPROVED** [review]

**Comment by anmyachev** [comment]
@devin-petersohn ready for review.

**Comment by vnlitvinov** [comment]
I've taken my fixes for `_inherit_docstrings` in a separate issue #3033 and fixing PR #3034, we should rebase this PR removing my commits when #3034 lands.

File: docs/flow/modin/base.rst
Comment (devin-petersohn) [inline docs/flow/modin/base.rst:1]: Should this be `modin/pandas/base.rst` instead?

File: docs/flow/modin/base.rst
Comment (devin-petersohn) [inline docs/flow/modin/base.rst:4]: I would clarify here that this is the pandas API.

**Review by devin-petersohn — COMMENTED** [review]
Thanks @gshimansky, left a couple of high level suggestions.

File: docs/flow/modin/base.rst
Comment (gshimansky) [reply docs/flow/modin/base.rst]: Ok.

File: docs/flow/modin/base.rst
Comment (gshimansky) [reply docs/flow/modin/base.rst]: Ok I added mention of pandas API, but if you still don't like it, please simply write that you want to see here or add your own commit. It is not easy to guess from a brief comment.

**Review by vnlitvinov — DISMISSED** [review]
LGTM with optional suggestions

File: docs/flow/modin/pandas/base.rst
Comment (devin-petersohn) [inline docs/flow/modin/pandas/base.rst:4]:
```suggestion
The class implements functionality that is common to Modin's pandas API for both ``DataFrame`` and ``Series`` classes.
```

File: docs/flow/modin/pandas/base.rst
Comment (devin-petersohn) [inline docs/flow/modin/pandas/base.rst:5]:
```suggestion
```

**Review by devin-petersohn — COMMENTED** [review]
Thanks for addressing the comments @gshimansky, left another rewording suggestion.

**Review by vnlitvinov — DISMISSED** [review]
LGTM

**Review by devin-petersohn — APPROVED** [review]
Thanks @gshimansky, LGTM!

**Pull request merged**
